cpp/src/arrow/csv/writer.cc

```diff
@@ -403,34 +415,44 @@ class CSVConverter {
   }
 
   static constexpr int64_t kColumnSizeGuess = 8;
+  io::OutputStream* sink_;
+  std::shared_ptr<io::OutputStream> owned_sink_;
```

Comment:
shared_ptr seems strange in general for a OutputStream which seems for the most part should have only one owner.

Reply:
I agree it seems weird, but both the IPC and Parquet writers use shared_ptr for this.

Reply:
Well, except that in Python any object can be shared, even if it's logically "owned" by something.